### PR TITLE
fix(iot-dev): Fix issue where amqp layer didn't recognize an ack to a subscription message that it sent

### PR DESCRIPTION
**Pull request by timtay-microsoft**
The recent twin array support changes introduced sending a desired properties subscription message each time the twin sender link opened. However, once the service acknowledges that message, our SDK logged an exception as it didn't correlate the ack back to the subscription message it sent. This PR adds a callback so that the top AMQP layer can be notified each time a lower level sends a subscription message

### CODE REVIEW

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 1]
```diff
@@ -688,6 +689,11 @@ else if (remoteState instanceof Modified || remoteState instanceof Released || r
                             this.listener.onMessageSent(inProgressMessages.remove(deliveryTag), transportException);
                         }
                     }
+                    else if (this.inProgressSubscriptionMessages.containsKey(deliveryTag))
+                    {
+                        SubscriptionType subscriptionType = this.inProgressSubscriptionMessages.remove(deliveryTag);
+                        log.debug("Successfully re-sent amqp subscription message of type {}", subscriptionType);
+                    }
                     else
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:697]: Without the above changes, this logic would fall to the else block and log an error

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java [hunk 2]
```diff
@@ -66,6 +66,7 @@
     private final static int MAX_MESSAGE_PAYLOAD_SIZE = 256*1024; //max IoT Hub message size is 256 kb, so amqp websocket layer should buffer at least that much space
     private final Boolean useWebSockets;
     private final Map<Integer, com.microsoft.azure.sdk.iot.device.Message> inProgressMessages = new ConcurrentHashMap<>();
+    private final Map<Integer, SubscriptionType> inProgressSubscriptionMessages = new ConcurrentHashMap<>();
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:69]: This new map will track outstanding subscription messages. I couldn't re-use the inProgressMessages map because there is no transport agnostic message to add to that map since we are creating the amqp message directly. Plus, messages in that map are expected to have a matching message in the IotHubTransport layer, which isn't the case for this special subscription message

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsDeviceOperations.java [hunk 3]
```diff
@@ -394,7 +394,7 @@ protected boolean onLinkRemoteOpen(String linkName)
         if (linkName.equals(this.getSenderLinkTag()))
         {
             this.amqpsSendLinkState = AmqpsDeviceOperationLinkState.OPENED;
-            this.log.debug("{} sender link with link correlation id {} was successfully opened {}", getLinkInstanceType(), this.linkCorrelationId);
+            this.log.debug("{} sender link with link correlation id {} was successfully opened", getLinkInstanceType(), this.linkCorrelationId);
```
Comment (timtay-microsoft) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsDeviceOperations.java:397]: This is orthogonal to the actual PR, but this log message had one too many spots for parameters, so I got rid of it

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionDeviceOperation.java [hunk 4]
```diff
@@ -29,14 +29,16 @@
 
     private List<UUID> cbsCorrelationIdList = Collections.synchronizedList(new ArrayList<UUID>());
 
+    private SubscriptionMessageRequestSentCallback subscriptionMessageRequestSentCallback;
+
     /**
      * Create logical device entity to handle all operation.
      *
      * @param deviceClientConfig the configuration of teh device.
      * @param amqpsDeviceAuthentication the authentication object associated with the device.
```
Comment (barustum) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionDeviceOperation.java:38]: Update constructor comments to reflect new parameter.

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/SubscriptionMessageRequestSentCallback.java [hunk 5]
```diff
@@ -0,0 +1,13 @@
+package com.microsoft.azure.sdk.iot.device.transport.amqps;
+
+import com.microsoft.azure.sdk.iot.device.DeviceTwin.DeviceOperations;
+
+public interface SubscriptionMessageRequestSentCallback
+{
+    void onSubscriptionMessageSent(int deliveryTag, SubscriptionType deviceOperationType);
```
Comment (azabbasi) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/SubscriptionMessageRequestSentCallback.java:7]: subscriptionType instead of deviceOperationType?

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionManager.java [hunk 6]
```diff
@@ -24,14 +24,15 @@
...
     private AmqpsDeviceAuthentication amqpsDeviceAuthentication;
     private ArrayList<AmqpsSessionDeviceOperation> amqpsDeviceSessionList = new ArrayList<>();
+    private SubscriptionMessageRequestSentCallback subscriptionMessageRequestSentCallback;
 
     /**
      * Constructor that takes a device configuration.
      *
      * @param deviceClientConfig the device configuration to use for 
      *                           session management.
      */
```
Comment (barustum) [inline device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionManager.java:34]: Update comment header to include new parameter

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionDeviceOperation.java (hunk 4 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsSessionDeviceOperation.java:38]: Good catch

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/SubscriptionMessageRequestSentCallback.java (hunk 5 above)
Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/SubscriptionMessageRequestSentCallback.java:7]: Good catch

File: device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java (hunk 1 above)
Comment (abhipsaMisra) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:697]: Why does the logs say: "re-sent amqp subscription message"?
I thought what we do here is -> if we receive an ack from service for the subscription message, we remove it from our internal queue, so as to not resend it. Is that not correct?

Comment (abhipsaMisra) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:697]: @timtay-microsoft : A question :raised_hand:

Comment (timtay-microsoft) [reply device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java:697]: This log statement gets executed when we receive an ack for an outstanding subscription message that we sent. It does get removed from the queue of outstanding messages as a result. The only odd thing about this subscription message is that it is sent at the session level. That's why this new queue needed to be added for messages like it

**Comment by timtay-microsoft** [comment]
I talked with Abhipsa about this, and she pointed out there is a small problem with this code. I'll get the fix checked in tomorrow. Don't merge this PR for now
